### DISCUSSION TITLE
IServerStartup "No parameterless constructor defined" trying to retrieve connectionstring

### DISCUSSION THREAD

**Discussion by artmedia** [oqtane/oqtane.framework#3379] [General]
Hello,
in a custom module I'm trying to retrieve connection string from IConfiguration in ServerStartup but I get this error: 
"Cannot dynamically create an instance of type 'Artmedia.MyProject.Repository.ServerStartup'. Reason: No parameterless constructor defined" 

I'm in localhost using ver.4.0.0
The code is this:

namespace Artmedia.MyProject.Repository
{
    public class ServerStartup : IServerStartup 
    {
        public IConfiguration _config { get; }
        public ServerStartup(IConfiguration configuration)
        {
            _config = configuration;
        }

        public void ConfigureServices(IServiceCollection services)
        {
            string _hangfireConnection = Utils.NormalizeConnectionString(_config.GetConnectionString("DefaultConnection"));
            // hangfire config using connctionstring...
        }
...

Why is it not possible? Or is there a better solution to get connctionstring?
Thanks
Luca

**Comment by sbwalker** [comment]
@artmedia it is impossible to determine from your code what you are trying to do. It is definitely possible to retrieve a connection string in Oqtane. However the approach would be different depending on whether your application is utilizing Oqtane's multi-tenant capabilities, is using EF Core, is using Dapper, is using standard ADO.NET, etc... Also I am not sure why you would need to use HangFire as Oqtane has its own background scheduled job processing capability already.

**Comment by artmedia** [comment]
Shaun, thanks for your response.
I'm trying to port into Oqtane as a custom module an old Asp.net application I wrote many years ago that sends Newsletters to a single subscriber every two seconds.
This operation takes minutes/hours if the email addresses are hundreds/thousands. Furthermore, for each submission I write an information in the database (sort of log), I white an information in database before the first sending and another after the last sending.
To achieve this I need an approach like FireAndForget that runs on a separate thread; HangFire should do it and it requires configuration of SqlServerStorage in ConfigureServices

Above explains what I'm doing, but **I'M NOT SURE if the error I'm getting is HangFire related**.
In fact, to get the error I simply need to add IConfiguration in ServerStartup in my Server project (without any HangFire instruction in ServerStartup):

public class ServerStartup : IServerStartup
{
public IConfiguration _config { get; }
public ServerStartup(IConfiguration configuration)
{
_config = configuration;
}
public void ConfigureServices(IServiceCollection services)
{
    string _myConnection = _config.GetConnectionString("DefaultConnection");
}
}

All my Oqtane instances, both localhost (dev) and production, are ver.4.0.0 single-tenant on IIS + SQL Server.
I'm NOT using EF Core, NOR using Dapper, only using standard ADO.NET

**Comment by sbwalker** [comment]
@artmedia I am not sure why you are trying to implement this within IServerStartup. IServerStartup is intended for registering services - not for doing any actual processing.  

Your code appears to be treating ServerStartup as a registered service as you are trying to pass in the Configuration service via its constructor - however ServerStartup is not a registered service at this point in the pipeline. You might have more luck if you dynamically retrieved the Configuration service from within the ConfigureServices() method:

var config = services.GetRequiredService< IConfiguration >();

However I am not even sure if that would work properly. Basically I don't think this is a valid approach. It would be better to create a DB Context class which can be registered in DI... and allow it to get the connection string. 

And in regards to Hangfire... Oqtane internally uses the .NET Core IHostedService interface which allows for background processing and FireAndForget, etc... in exactly the same way as Hangfire - except without having to manage a third party dependency.

**Comment by artmedia** [comment]
I didn't think I could use the IHostedService interface for the FireAndForget approach.
Thank you for your explanation.

**Comment by sbwalker** [comment]
@artmedia I should probably clarify.... Oqtane has scheduled jobs which utilize the IHostedService interface but wrap additional services around it so that you can schedule jobs to run on a specific schedule and maintain a log of all job executions, etc... However the lowest level of time granularity for scheduled jobs is 1 minute. So if you truly want to send emails every 2 seconds then you may need to implement your own custom IHostedService so that you can implement your own timer, etc... This is not difficult - there are plenty of examples/documentation available from Microsoft.
